image_get/worker/test/test_worker.py

```diff
@@ -0,0 +1,219 @@
+import json
+import pytest
+import asyncio
+import time
+import random
+import logging as log
+from worker.consumer import poll_consumer, consume
+from worker.util import process_image
+from PIL import Image
+from collections import deque
+from enum import Enum, auto
+from functools import partial
+
+log.basicConfig(level=log.DEBUG)
+
+
+class FakeMessage:
+    def __init__(self, value):
+        self.value = value
+
+
+class FakeConsumer:
+    def __init__(self):
+        self.messages = []
+
+    def insert(self, message):
+        self.messages.append(
+            FakeMessage(bytes(message, 'utf-8'))
+        )
+
+    def consume(self, block=True):
+        if self.messages:
+            return self.messages.pop()
+        else:
+            return None
+
+    def commit_offsets(self):
+        pass
+
+
+class FakeImageResponse:
+    def __init__(self, status=200, corrupt=False):
+        self.status = status
+        self.corrupt = False
+
+    async def read(self):
+        # 1024 x 768 sample image
+        if self.corrupt:
+            location = 'test/corrupt.jpg'
+        else:
+            location = 'test/test_image.jpg'
+        with open(location, 'rb') as f:
+            return f.read()
+
+
+class FakeAioSession:
+    def __init__(self, corrupt=False):
+        self.corrupt = corrupt
+
+    async def get(self, url):
+        return FakeImageResponse(self.corrupt)
+
+
+class AioNetworkSimulatingSession:
+    """
+    It's a FakeAIOSession, but it can simulate network latency, errors,
+    and congestion. At 80% of its max load, it will start to slow down and occasionally
+    throw an error. At 100%, error rates become very high and response times slow.
+    """
```

Comment:
Sorry about the import issue, there was a PYTHONPATH-related issue that I've fixed now.

As for the rate limits, that's addressed in https://github.com/creativecommons/cccatalog-api/issues/466 and in `rate_limit.py`; we can back off if there are a lot of error messages or degredation of time-to-first-byte. The plan is to learn the ideal rate limit for each upstream source through automated testing.

Reply:
Also, just to be clear, that's a simulator for testing purposes, not our actual rate limiting implementation

Reply:
Yes, for the most part.  I think it's a good strategy for the many providers we don't want to spend any individual effort on for this.  I was thinking that it might make some sense to get a specific feel for how the Flickr and WMC providers will act and write tests that emulate their specific behavior, since they'll be providing most images (and therefore optimizing for scraping from them will have the most benefit).

Reply:
I can make some adjustments to the network simulator to fail in different ways once I make some more progress on the rate limiting implementation. In any case, we'll have to monitor the progress of the crawl in real-world conditions to make sure that the approach is working as intended.